updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Result;
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "updatehub-statechange.sock";
+
+type StateChangeCallback = fn(State);
+type ErrorCallback = fn(String);
+
+pub struct StateChangeListener {
+    listeners: HashMap<String, StateChangeCallback>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    id: StateID,
+    conn: UnixStream,
+}
+
+impl State {
+    pub fn downloading(&mut self) {
+        self.conn.write_all(b"downloading");
+    }
+
+    pub fn rebooting(&mut self) {
+        self.conn.write_all(b"rebooting");
+    }
+
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed() {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on(&mut self, state: StateID, f: fn(State)) {
+        let name = match state {
+            StateID::Download => String::from("download"),
+            StateID::Install => String::from("install"),
+            StateID::Reboot => String::from("reboot"),
+            StateID::Error => String::from("error"),
+        };
+        self.listeners.insert(name, f);
+    }
+
+    fn on_error(&mut self, f: fn(String)) {
+        self.error_handlers.push(f);
+    }
+
+    fn emit(&self, conn: UnixStream, state: &str) {
+        self.listeners.iter().for_each(|(_, f)| {
+            let conn = conn.clone();
+            f(State {
+                id: match state {
+                    "download" => StateID::Download,
+                    "install" => StateID::Install,
+                    "reboot" => StateID::Reboot,
+                    "error" => StateID::Error,
+                    _ => panic!("error"),
+                },
+                conn,
+            })
+        });
+    }
+
+    fn throw_error(&self, error: &str) {
+        self.error_handlers.iter().for_each(|f| {
+            let error = &error;
+            f(error.to_string())
+        })
+    }
+
+    async fn listen(&self) -> Result<()> {
+        let file = std::path::Path::new(&SDK_TRIGGER_FILENAME);
+        let listener = if file.exists() {
+            UnixListener::bind(format!("/run/{}", SOCKET_PATH)).await?
+        } else {
+            println!(
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}",
+                SDK_TRIGGER_FILENAME
+            );
+            UnixListener::bind(SOCKET_PATH).await?
```

Comment:
Socket should always be opened to the default path on `/run/updatehub-statechange.sock` unless a env variable changes it, so we have this implementation compatible with the others SDK.